design/ceph-deprecate-old-osd-types.md

```diff
@@ -0,0 +1,433 @@
+# Ceph OSD redesign for Ceph Octopus (May 2019)
+**Targeted for v1.1-v1.2**
+
+## Background
+Rook's current implementation for preparing and starting OSDs includes several different patterns
+that have evolved over Rook's lifetime. These patterns include different code paths for supporting
+directories and devices as storage targets, and further separated into support for both Filestore
+and Bluestore as backing stores. There is further complication added by support for `ceph-volume` as
+a tool for storage provisioning versus continued support for legacy installations of Rook which used
+manual partitioning.
+
+There has been talk for quite some time about removing support for directories as storage targets,
+as this is a use case in Ceph that is not supported for production. Additionally, Filestore is
+effectively in sustaining mode, and Bluestore is the defacto replacement for it.
+
+All of the various OSD support patterns/code-paths need changes to allow the [planned updates to
+Rook's configuration strategy](/design/ceph-config-updates.md) to proceed. The various code paths
+are intertangled and difficult to separate, and this presents a huge hurdle for the config changes.
+In general, the intertangled OSD code poses a risk for long-term maintenance, as it may be more
+prone to bugs being introduced, and it has and will continue to slow development.
+
+Atop all of this, there have been discussions ongoing about supporting Ceph's Rook
+orchestrator as the intended way of managing Ceph deployments on Kubernetes. Orchestration currently
+modifies the Ceph cluster resource, but from the standpoint of a Ceph storage administrator, the CRD
+(and Rook) lacks the flexibility needed for non-trivial Ceph deployments. While flexibility could be
+added, it will be better to remove the need for such complexity in the cluster CRD and in Rook. The
+details of this part of the design will be expounded later.
+
+In light of this background information, this design proposes three high-level goals:
+1. Identify OSD features/code-paths that should be deprecated due to having become outmoded in some
+   form.
+2. Propose a new architecture for Rook's OSD code that is compatible with the planned config updates
+   and that is extensible for the future.
+3. Provide a timeline for deprecating outmoded, and provide users tools for transitioning their
+   deployments which use old methods to begin using the new methods.
+
+
+## Deprecated features and code paths
+This design proposes to deprecate ...
+- the use of directories as storage targets
+- support for the Filestore backing store
+- support for devices not provisioned with `ceph-volume`
```

Comment:
I'm not sure we need to deprecate old-fashioned prepared OSDs. Just like the transition between `ceph-disk` and `ceph-volume`, OSDs continue to operate. So we should be able to *detect* and *activate* these OSDs but not allow the preparation (already done since `ceph-volume` is the default.

Reply:
As clarity, this design proposes to deprecate these but not after providing users a means of migrating from old to new.